File: dbscan.js
//    Y
//    /\
//    |
// 8  |
// 7  |   p0(2,6)                          p6(12,6)
// 6  |       *                             *
// 5  |   p1(2,4) p3(4,4) p4(7,4) p5(10,4) p7(12,4)
// 4  |       *     *        *        *     *  
// 3  |   p2(2,2)                          p8(12,2)
// 2  |       *                             *
// 1  |
// 0  -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -----> X
//      0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15

const db = [{
  id: 0,
  label: undefined,
  pos: {
    x: 2,
    y: 6
  }
}, {
  id: 1,
  label: undefined,
  pos: {
    x: 2,
    y: 4
  }
}, {
  id: 2,
  label: undefined,
  pos: {
    x: 2,
    y: 2
  }
}, {
  id: 3,
  label: undefined,
  pos: {
    x: 4,
    y: 4
  }
}, {
  id: 4,
  label: undefined,
  pos: {
    x: 7,
    y: 4
  }
}, {
  id: 5,
  label: undefined,
  pos: {
    x: 10,
    y: 4
  }
}, {
  id: 6,
  label: undefined,
  pos: {
    x: 12,
    y: 6
  }
}, {
  id: 7,
  label: undefined,
  pos: {
    x: 12,
    y: 4
  }
}, {
  id: 8,
  label: undefined,
  pos: {
    x: 12,
    y: 2
  }
}];

function dist(p1, p2) {
  let res = Math.sqrt(
    Math.pow(p2.pos.x - p1.pos.x, 2) +
    Math.pow(p2.pos.y - p1.pos.y, 2)
  );
  console.log('===>' + res);
  return res;
}

function rangeQuery(db, q, eps) {
  let neighbors = [];
  db.map(p => { // Scan all points in the database
    if (dist(p, q) <= eps) { // Compute distance and check epsilon
      neighbors.push(p); // Add to result
    }
  });
  return neighbors;
}

function DBSCAN(db, eps, minPts) {
  let c = 0; // Cluster counter
  db.map(p => {
    if (p.label) {
      return;
    } // Previously processed in inner loop
    let neighbors = rangeQuery(db, p, eps); // Find neighbors
    if (neighbors.length < minPts) { // Density check
      p.label = 'Noise'; // Label as Noise
      return;
    }
    c++; // next cluster label
    p.label = c; // Label initial point
    let seeds = neighbors.filter(n => n.id !== p.id); // Neighbors to expand
    seeds.map(s => { // Process every seed point
      if (s.label == 'Noise') {
        s.label = c;
      } // Change Noise to border point
      if (s.label) {
        return;
      } // Previously processed
      s.label = c; // Label neighbor
      let seedNeighbors = rangeQuery(db, s, eps); // Find neighbors
      if (seedNeighbors.length >= minPts) { // Density check
        seeds.push(...seedNeighbors); // Add new neighbors to seed set
      }
    });
  });
}

const eps = 2.5;
const minPts = 4;

DBSCAN(db, eps, minPts);
console.log(db);

File: dbscan.go
package main


import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"math"
)

type Point struct {
	Id		int
	Label	string
	X		int `json:"x"`
	Y 		int `json:"y"`
}

func Read(file string) []Point {
	jsonFile, err := os.Open(file)
	if err != nil {
		fmt.Println(err)
		return nil
	} else {
		byteValue, _ := ioutil.ReadAll(jsonFile)
		var points []Point
		json.Unmarshal([]byte(byteValue), &points)
		defer jsonFile.Close()
		return points
	}
}

func Init(points *[]Point) {
	for i:=0; i < len(*points); i++ {
		(*points)[i].Id = i;
		(*points)[i].Label = "undefined";		
	}
}

func Dist(p1, p2 Point) float64 {
	var res float64 = math.Sqrt(math.Pow(float64(p2.X - p1.X), 2) + math.Pow(float64(p2.Y - p1.Y), 2));
	return res;
}

func RangeQuery(db *[]Point, q Point, eps float64) []Point {
	var neighbors []Point
	for i := 0; i < len(*db); i++ {
		if (Dist((*db)[i], q) <= eps) {
			neighbors = append(neighbors, (*db)[i])
		}
	}
	return neighbors;
}

func DBSCAN(db *[]Point, eps float64, minPts int) {
	var c int = 0; // Cluster counter
	for i := 0; i < len(*db); i++ {
		if (*db)[i].Label != "undefined" {
			continue
		} // Previously processed in inner loop
		var neighbors = RangeQuery(db, (*db)[i], eps); // Find neighbors
		if (len(neighbors) < minPts) { // Density check
			(*db)[i].Label = "Noise" // Label as Noise
			continue
		}
		c++ // next cluster label
		(*db)[i].Label = strconv.Itoa(c) // Label initial point
		var seeds = neighbors // .filter(n => n.id !== p.id); // Neighbors to expand
		for j := 0; j < len(seeds); j++ { // Process every seed point
			if (seeds[j].Id == (*db)[i].Id) {
				continue
			}
			if (seeds[j].Label == "Noise") {
				seeds[j].Label = strconv.Itoa(c)
			} // Change Noise to border point
			if (seeds[j].Label != "undefined") {
				continue
			} // Previously processed
			seeds[j].Label = strconv.Itoa(c); // Label neighbor
			var seedNeighbors = RangeQuery(db, seeds[j], eps); // Find neighbors
			if (len(seedNeighbors) >= minPts) { // Density check
				for k := 0; k < len(seedNeighbors); k++ {
					seeds = append(seeds, seedNeighbors[k]); // Add new neighbors to seed set
				}
				fmt.Println("new len" + strconv.Itoa(len(seeds)))
			}
		}
	}

	fmt.Println("Done")

}

func main() {
	
	var points []Point = Read("input1.json")
	pointer := &points

	Init(&points)

	var eps float64 = 2.5
	var minPts int = 4
	DBSCAN(&points, eps, minPts)

	for i := 0; i < len(*pointer); i++ {
		fmt.Println("Id: " + strconv.Itoa((*pointer)[i].Id)) // strconv.Itoa(points[i].X))
		fmt.Println("Label: " + (*pointer)[i].Label) //strconv.Itoa(points[i].Y))
	}

	fmt.Println("Successfully Opened data.json")
	// defer the closing of our jsonFile so that we can parse it later on
	
}